--- a/src/GestureManager.cpp
+++ b/src/GestureManager.cpp
@@ -11,6 +11,7 @@
 #include <hyprland/src/config/ConfigValue.hpp>
 #include <hyprland/src/config/legacy/ConfigManager.hpp>
 #include <hyprland/src/desktop/state/FocusState.hpp>
+#include <hyprland/src/managers/fullscreen/FullscreenController.hpp>
 #include <hyprland/src/output/Monitor.hpp>
 #include <hyprland/src/pointer/PointerController.hpp>
 #include <hyprland/src/state/MonitorState.hpp>
@@ -187,10 +188,10 @@
                 const Vector2D touchPos =
                     pixelPositionToPercentagePosition(this->m_sGestureState.get_center().current) *
                     this->m_lastTouchedMonitor->m_size;
-                if (w && !w->isFullscreen()) {
+                if (w && !Fullscreen::controller()->isFullscreen(w)) {
                     const CBox real = {
-                        w->m_realPosition->value().x, w->m_realPosition->value().y, w->m_realSize->value().x,
-                        w->m_realSize->value().y
+                        w->positionAnimation()->value().x, w->positionAnimation()->value().y, w->sizeAnimation()->value().x,
+                        w->sizeAnimation()->value().y
                     };
                     const CBox grab = {
                         real.x - BORDER_GRAB_AREA, real.y - BORDER_GRAB_AREA, real.width + 2 * BORDER_GRAB_AREA,
@@ -198,7 +199,7 @@
                     };

                     bool notInRealWindow = !real.containsPoint(touchPos) || w->isInCurvedCorner(touchPos.x, touchPos.y);
-                    bool onTiledGap      = !w->m_isFloating && !w->isFullscreen() && notInRealWindow;
+                    bool onTiledGap      = !w->m_isFloating && !Fullscreen::controller()->isFullscreen(w) && notInRealWindow;
                     bool inGrabArea      = notInRealWindow && grab.containsPoint(touchPos);

                     if ((onTiledGap || inGrabArea) && !w->hasPopupAt(touchPos)) {
